Add API to access welcome content

dcermak opened the pull request:
This is a very crude and far from finished PR to add support for [Welcome Content](https://code.visualstudio.com/api/extension-guides/tree-view#welcome-content) in tree views.

To make things a bit complicated, every tree view section has a `welcome-view` and `welcome-view-content` sub element, but they are invisible/without content. When they are actually visible, then the content is present in the div with the class name `pane-body welcome`.

The part that is currently missing is which content to get out of the welcome view. Welcome views have afaik only text and buttons which can be clicked. The content is present as child element of the div with class name `welcome-view-content`. Text content is simply present in `<p>` elements, the buttons on the other hand use the `a` element with the class `monaco-button monaco-text-button`.

I would propose to create a class `WelcomeContent` that is returned by `getWelcomeContent()` and that has the function `getContent()` which would return a Promise returning an Array of either text (extracted from `<p></p>`) or actionable buttons (these could be similar to the way notification buttons behave).

dcermak commented:
I have added a draft of such a possible welcome content class & extended the tests.

Please tell me if this is something that you'd like and how the API should look, so that I can polish it a bit more and add more documentation.

jrichter1 commented on page-objects/src/components/sidebar/WelcomeContent.ts:
```diff
@@ -0,0 +1,40 @@
+  }
+
+  public async getTitle(): Promise<string> {
+    return this.panel.getText();
```
`this.getText()` should suffice

jrichter1 commented on page-objects/src/components/sidebar/WelcomeContent.ts:
```diff
@@ -0,0 +1,40 @@
+export class WelcomeContentButton extends AbstractElement {
+  constructor(
+    private panel: WebElement,
+    public welcomeSection: WelcomeContentSection
```
You don't really need to declare the arguments as fields here.
The first (`panel`) argument is the main reference for selenium, after the object is initialized, `this` is populated with whatever element is found by this reference.
The second (`welcomeSection`) argument is accessible as a protected field `enclosingElement`, or in `getEnclosingElement()` method publicly.

jrichter1 commented on page-objects/src/components/sidebar/WelcomeContent.ts:
```diff
@@ -0,0 +1,40 @@
+    return this.panel.getText();
+  }
+
+  public async takeAction(): Promise<void> {
```
looks easier to just call `click` instead, no need for this method :)

dcermak replied:
Indeed. That makes me wonder whether this class is actually even required, because `takeAction()` can be replaced with `button.click()` and `getTitle()`  with `button.getText()`, leaving the `WelcomeContentButton` only with the constructor. Not sure if it is worth having a separate class for that? Maybe just a typedef instead?

jrichter1 commented on page-objects/src/components/sidebar/WelcomeContent.ts:
```diff
@@ -0,0 +1,40 @@
+
+export class WelcomeContentSection extends AbstractElement {
+  constructor(public parent: ViewSection, panel: WebElement) {
+    super(panel);
```
For the sake of consistence, please use the parent element as the second argument. Also, looks like you forgot to feed the parent into the super constructor.

dcermak commented on page-objects/src/components/sidebar/ViewSection.ts:
```diff
@@ -60,6 +61,20 @@ export abstract class ViewSection extends AbstractElement {
+        }
+    }
+
+    public async getWelcomeContent(): Promise<WelcomeContentSection> {
```
That sounds like a good idea, I'll change that.

dcermak commented:
I have pushed an updated version of the proposed API and polished it a bit more.

I have run into one tiny issue: welcome content was added in vscode 1.44, but vscode-extension-tester supports vscode down to 1.37. How should this be addressed, especially wrt to the locators? Should I mark the welcome content locator as `undefined` by default and make the welcome content find function return undefined if the locator is undefined?

jrichter1 commented:
You can just dump all the new locators into 1.37.0 without a problem. There is no harm in having a superset of locators for a particular version, some of them simply won't be used. It's not like we could prevent the users from using the APIs meant for newer versions of vscode anyway. The best we can do here is put the version requirement in the docs.

Btw. the rest of the code looks pretty good.

PS: I don't really actively support more than 5 latest versions of vscode anymore.

dcermak commented:
Jan Richter <notifications@github.com> writes:

> You can just dump all the new locators into 1.37.0 without a problem. There is no harm in having a superset of locators for a particular version, some of them simply won't be used. It's not like we could prevent the users from using the APIs meant for newer versions of vscode anyway. The best we can do here is put the version requirement in the docs.

I have removed the WIP comment from that file then.

>
> Btw. the rest of the code looks pretty good.

Thanks! I've extended the documentation a bit and removed the WIP from
the commit message.

jrichter1 commented on page-objects/src/components/sidebar/WelcomeContent.ts:
```diff
@@ -0,0 +1,80 @@
+     * Combination of [[getButtons]] and [[getTextSections]]: returns all entries in the welcome view in the order that they appear.
+     */
+    public async getContents(): Promise<(WelcomeContentButton|string)[]> {
+        const elements = await this.findElements(By.xpath(".//a[@class='monaco-button monaco-text-button'] | .//p"));
```
please externalize all the locators

dcermak replied:
Done.

jrichter1 commented on page-objects/src/components/sidebar/WelcomeContent.ts:
```diff
@@ -0,0 +1,80 @@
+            } else if (tagName === "a") {
+                return new WelcomeContentButton(e, this);
+            } else {
+                assert(false, `got an invalid tag name: '${tagName}', expected 'p' or 'a'`);
```
wouldn't throwing an error be simpler?

dcermak replied:
Well, kinda. This is more of a "philosophical" question though: my reasoning here is that we expect only buttons and text and anything else means we have a bug in our code (thus we assert that this does not happen). An exception on the other hand means that a user can reach this code path due to an error. But this is something that **must** not happen.

However, that's just my stance on this, so feel free to disagree and I'll change it to throw or remove the else branch entirely.

jrichter1 replied:
From the API standpoint, `assert(false)` is just a way to throw an `AssertionError`. And I'd say these should be reserved for failing test conditions. The page objects are not really a test framework (or an assertion framework either).

dcermak replied:
I have removed the current else branch entirely and moved the else if into the else branch.

jrichter1 reviewed (approved):
Looking good

dcermak commented:
Thanks!
